provider/datagen/src/driver.rs

```diff
@@ -5,20 +5,265 @@
+    Maximal,
+    #[allow(dead_code)]
+    RetainBaseLanguages,
+    NoDeduplication,
+}
+
+impl DeduplicationStrategy {
+    fn describe(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let s = match self {
+            Self::Maximal => "maximal deduplication",
+            Self::RetainBaseLanguages => "deduplication retaining base languages",
+            Self::NoDeduplication => "no deduplication",
+        };
+        write!(f, "{s}")
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct LocaleWithExpansion {
+    langid: LanguageIdentifier,
+    include_ancestors: bool,
+    include_descendants: bool,
+}
+
+impl LocaleWithExpansion {
+    // en-US
+    pub fn with_variants(langid: LanguageIdentifier) -> Self {
+        Self {
+            langid,
+            include_ancestors: true,
+            include_descendants: true,
+        }
+    }
+
+    // ^en-US
+    #[allow(dead_code)]
+    pub fn without_variants(langid: LanguageIdentifier) -> Self {
+        Self {
+            langid,
+            include_ancestors: true,
+            include_descendants: false,
+        }
+    }
+
+    // @en-US
+    #[allow(dead_code)]
+    pub fn preresolved(langid: LanguageIdentifier) -> Self {
+        Self {
+            langid,
+            include_ancestors: false,
+            include_descendants: false,
+        }
+    }
+
+    pub(crate) fn into_langid(self) -> LanguageIdentifier {
+        self.langid
+    }
+
+    pub(crate) fn wildcard() -> Self {
+        Self {
+            langid: langid!("und"),
+            include_ancestors: true,
+            include_descendants: true,
+        }
+    }
+}
+
+impl fmt::Display for LocaleWithExpansion {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if self.include_descendants {
+            write!(f, "{}", self.langid)
+        } else if self.include_ancestors {
+            write!(f, "^{}", self.langid)
+        } else {
+            write!(f, "@{}", self.langid)
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub struct LocalesWithFallback {
+    pub runtime_fallback_location: Option<RuntimeFallbackLocation>,
+    pub deduplication_strategy: Option<DeduplicationStrategy>,
+    /// private: set via constructor
+    locales: HashSet<LocaleWithExpansion>,
+}
+
+impl LocalesWithFallback {
+    #[inline]
+    #[allow(dead_code)]
+    pub fn for_locales(locales: impl IntoIterator<Item = LocaleWithExpansion>) -> Self {
+        Self::from_iter(locales)
+    }
+
+    pub fn for_all_locales() -> Self {
+        Self::from_iter([LocaleWithExpansion::wildcard()])
+    }
+
+    fn describe(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut sorted_locales = self
+            .locales
+            .iter()
+            .map(ToString::to_string)
+            .collect::<Vec<_>>();
+        sorted_locales.sort();
+        let start = match self.runtime_fallback_location {
+            None => "with fallback",
```

Comment:
I think this should be unreachable. We should only log this after we resolved against the exporter

Reply:
done